### ISSUE TITLE
[Discussion] Adding pre-commit hooks

### ISSUE THREAD

**Issue by asvetlov** [aio-libs/aiohttp#5147]
I'd like to discuss the tool. 
TLDR: should we use `pre-commit`, and how?

Today I've added https://github.com/asottile/pyupgrade tool to fix some obsolete syntax in our code-base.
The tool works like a charm, but I did not read the usage instruction well and installed it not as a standalone script but as a part of https://pre-commit.com/ hooks.
`pre-commit` also works pretty well and can be considered as the single configuration for both *linters* and *fixers*.

The only problem is `pre-commit` ideology. It applies all hooks enumerated in pre-commit configuration and fails if either hook explicitly fails or it modified some files.
E.g. I have a modified python file that should be fixed by `black`. I do `git add ...` and `git commit`.
On the `commit` the hook is executed. It fails, shows an error, **and** reformats the source file by fixing it.
Maybe fixing. Sometimes the next run fails as well and you should try several times.
Not bad but I did not grok the behavior instantly.  But, well, handy if you know how the tool works.

Small note: by default `pre-commit` runs hooks only over **committed** files which sounds a good compromise between the check performance and quality. The only exception that I know is `mypy` checker which requires the whole source tree to analyze (yes, I know that mypy can be applied to explicit files only but this mode detects much less typing errors). 

By default, you implicitly run `pre-commit` every time you type `git commit` command after setting the tool up (`pre-commit install`) once.
There is an option to run `pre-commit` over all files explicitly by `pre-commit run --all-files`.  I can add `make` aliases for it; `make fmt` checks all tools and `make lint` reformats all source as a side effect.

I see three options:
1. Add `pre-commit`, drop `make fmt`, `make lint`, and all other linting commands
2. Add `pre-commit`, convert `make fmt` / `make lint` to aliases for `pre-commit run --all-files`, drop other linting commands from `Makefile`.
3. Don't use `pre-commit`, tidy up `make fmt` / `make lint` to do its job well, drop all other lint commands like `make flake` and `make isort`.

I'm inclining to option 2.

All @aio-libs/aiohttp-committers @aio-libs/admins @aio-libs/triagers and just contributors who are not members of the project (yet) -- I'd like to know your opinion.

P.S.
As you see, I want to clean up Makefile anyway.
P.P.S.
@webknjaz I know I can use GitHub Discussions for such things but please let me use the Issues. I have a feeling that the Discussions is a little less visible to contributors; I want to hear as large a community as possible.

**Comment by webknjaz** [comment]
I've been using pre-commit for 5+ years — it's a defacto default linter collection wrapper these days so I don't see any problem with it (here's an example of how I sometimes set it up: https://github.com/ansible/pylibssh/blob/devel/.pre-commit-config.yaml). It's usually set up via tox, though (another defacto standard tool) — I cannot judge what you change in the make-oriented setup because I never use it (since I don't like the idea of something randomly mutating my env). Also, aiohttp is almost the only project that I'm involved in not using tox. I have no comments on black either because I hate the idea of reformatting perfectly good code into something less readable.

By the way, you're not forced to install it as a pre-commit hook. It's less known but some folks would use it on pre-push. The most important is to run it in CI, local env is optional.

re:discussions — they are still in private beta so nobody is obliged to use them. I think over time they'll become better as GH folks iterate over improvements and address the feedback. Also, if you have any feedback — let me know and I'll make sure GH people will hear it.

**Comment by mjpieters** [comment]
I've been a very happy user of pre-commit for a long time now, and would definitely welcome it's use.

The only downside I can see is that dependabot doesn't know about `.pre-commit-hooks.yml` and won't suggest updates to the hook versions. We could run `pre-commit autoupdate; pre-commit run --all` periodically to verify that hooks can and should be updated.

**Comment by asvetlov** [comment]
I thought about github action that runs `pre-commit autoupdate` and fails if something is actually updated.
Making a PR is even better but requires more effort.
A hero for github workflow PR is welcome!

**Comment by asvetlov** [comment]
https://browniebroke.com/blog/gh-action-pre-commit-autoupdate/ article describes a solution for auto-updating

**Comment by asvetlov** [comment]
Done
